cli/tools/jupyter/server.rs

```diff
@@ -0,0 +1,714 @@
+              "cursor_end": cursor_pos,
+              "metadata": {},
+            }))
+            .send(connection)
+            .await?;
+        }
+      }
+      "comm_msg" | "comm_info_request" | "history_request" => {
+        // We don't handle these messages
+      }
+      _ => {
+        eprintln!("Unrecognized shell message type: {}", msg.message_type());
+      }
+    }
+
+    msg
+      .new_message("status")
+      .with_content(json!({"execution_state": "idle"}))
+      .send(&mut *self.iopub_socket.lock().await)
+      .await?;
+    Ok(())
+  }
+
+  async fn handle_execution_request(
+    &mut self,
+    msg: JupyterMessage,
+    connection: &mut Connection<zeromq::RouterSocket>,
+  ) -> Result<(), AnyError> {
+    self.execution_count += 1;
+    *self.last_execution_request.borrow_mut() = Some(msg.clone());
+
+    msg
+      .new_message("execute_input")
+      .with_content(json!({
+          "execution_count": self.execution_count,
+          "code": msg.code()
+      }))
+      .send(&mut *self.iopub_socket.lock().await)
+      .await?;
+
+    let result = self
+      .repl_session
+      .evaluate_line_with_object_wrapping(msg.code())
+      .await;
+
+    let evaluate_response = match result {
+      Ok(eval_response) => eval_response,
+      Err(err) => {
+        msg
+          .new_message("error")
+          .with_content(json!({
+            "ename": err.to_string(),
+            "evalue": "",
+            "traceback": [],
+          }))
+          .send(&mut *self.iopub_socket.lock().await)
+          .await?;
+        msg
+          .new_reply()
+          .with_content(json!({
+            "status": "error",
+            "execution_count": self.execution_count,
+          }))
+          .send(connection)
+          .await?;
+        return Ok(());
+      }
+    };
+
+    let repl::cdp::EvaluateResponse {
+      result,
+      exception_details,
+    } = evaluate_response.value;
+
+    if exception_details.is_none() {
+      let output =
+        get_jupyter_display_or_eval_value(&mut self.repl_session, &result)
+          .await?;
+      msg
+        .new_message("execute_result")
+        .with_content(json!({
+            "execution_count": self.execution_count,
+            "data": output,
+            "metadata": {},
+        }))
+        .send(&mut *self.iopub_socket.lock().await)
+        .await?;
+      msg
+        .new_reply()
+        .with_content(json!({
+            "status": "ok",
+            "execution_count": self.execution_count,
+        }))
+        .send(connection)
+        .await?;
+      // Let's sleep here for a few ms, so we give a chance to the task that is
+      // handling stdout and stderr streams to receive and flush the content.
+      // Otherwise, executing multiple cells one-by-one might lead to output
+      // from various cells be grouped together in another cell result.
+      tokio::time::sleep(std::time::Duration::from_millis(5)).await;
```

Comment:
This seems a little fragile but maybe we can add a TODO to replace it with a mutex or other lock?

Reply:
There's nothing to mutex/lock on - it's the fact that channels need a moment to flush all data that might have been printed/console logged. I haven't found a better solution to this problem.

Reply:
Fair enough. Looks like the protocol for the kernels doesn't seem to give us any way to actually route these properly. We can always adjust in the future if the heuristic doesn't work.

Reply:
FWIW it seems like ipython's kernel does the same thing

https://github.com/ipython/ipykernel/blob/83ab7d02958f8134ddabf9d2722c222eaab041d4/ipykernel/kernelbase.py#L202

Reply:
Protocol wise this can be handled properly by sending the right parent_header.msg_id. However that would mean tracking the parent message ID for each execution which might not be in the right cell during async code.

On the Jupyter front ends we determine all the outputs for a cell by grouping on the parent--header.msg_id